src/3d/qgstessellatedpolygongeometry.cpp

```diff
@@ -123,16 +184,29 @@ void QgsTessellatedPolygonGeometry::setData( const QByteArray &vertexBufferData,
     mTextureCoordsAttribute->setCount( vertexCount );
 }
 
+bool QgsTessellatedPolygonGeometry::rayIntersection( const QgsRayCastingUtils::Ray3D &ray, const QMatrix4x4 &worldTransform, QVector3D &intersectionPoint, QgsFeatureId &fid )
+{
+  int triangleIndex = -1;
+  const int stride = static_cast<int>( mPositionAttribute->byteStride() );
+  bool success = intersectionTriangles( mVertexBuffer->data(), stride, ray, worldTransform, intersectionPoint, triangleIndex );
 
-// run binary search on a sorted array, return index i where data[i] <= x < data[i+1]
+  fid = success ? triangleIndexToFeatureId( triangleIndex ) : FID_NULL;
+
+  return success;
+}
+
+// run binary search on a sorted array, return index i where data[i] <= v < data[i+1]
 static int binary_search( uint v, const uint *data, int count )
 {
   int idx0 = 0;
   int idx1 = count - 1;
 
-  if ( v < data[0] || v >= data[count - 1] )
+  if ( v < data[0] )
     return -1;  // not in the array
 
+  if ( v >= data[count - 1] )
```

Comment:
Is it safe? Is count -1 will not be out of range ?

Reply:
count is data's size and will never be called with count < 1